mino/minows/session.go

```diff
@@ -0,0 +1,250 @@
+	}
+	o.logger.Debug().Stringer("to", dest).Msgf("relayed packet")
+	return nil
+}
+
+func (o orchestrator) listen(decoder *gob.Decoder) (Packet, error) {
+	for {
+		packet, dest, err := o.fetch(decoder)
+		if err != nil {
+			return Packet{}, xerrors.Errorf("could not receive: %v", err)
+		}
+		switch to := dest.(type) {
+		case orchestratorAddr:
+			if o.myAddr.Equal(to) {
+				return packet, nil
+			}
+		case address:
+			err := o.relay(packet, to)
+			if err != nil {
+				return Packet{}, xerrors.Errorf("could not relay: %v", err)
+			}
+		}
+	}
+}
+
+type participant struct {
+	logger zerolog.Logger
+
+	myAddr address
+	rpc    rpc
+	// Connects to the orchestrator
+	out *gob.Encoder
+	in  chan Packet
+}
+
+func (p participant) Send(msg serde.Message, addrs ...mino.Address) <-chan error {
+	return doSend(addrs, msg, p.send, p.logger)
+}
+
+func (p participant) Recv(ctx context.Context) (mino.Address, serde.Message, error) {
+	unpack := unpacker(p.rpc.mino.GetAddressFactory(), p.rpc.factory, p.rpc.context)
+	return doReceive(ctx, p.in, unpack, p.logger)
+}
+
+func (p participant) send(addr mino.Address, msg serde.Message) error {
+	switch addr.(type) {
+	case address:
+	case orchestratorAddr:
+	default:
+		return xerrors.Errorf("%v: %T", ErrWrongAddressType, addr)
+	}
+
+	src, err := p.myAddr.MarshalText()
+	if err != nil {
+		return xerrors.Errorf("could not marshal address: %v", err)
+	}
+	payload, err := msg.Serialize(p.rpc.context)
+	if err != nil {
+		return xerrors.Errorf("could not serialize message: %v", err)
+	}
+	dest, err := addr.MarshalText()
+	if err != nil {
+		return xerrors.Errorf("could not marshal address: %v", err)
+	}
+
+	// Send to orchestrator to relay to the destination participant
+	forward := Forward{
+		Packet:      Packet{Source: src, Payload: payload},
+		Destination: dest,
+	}
+	err = p.out.Encode(&forward)
+	if err != nil {
+		return xerrors.Errorf("could not encode packet: %v", err)
+	}
+	return nil
+}
+
+func (p participant) listen(decoder *gob.Decoder) (Packet, error) {
+	var packet Packet
+	err := decoder.Decode(&packet)
+	if err != nil {
+		return Packet{}, xerrors.Errorf("could not decode packet: %v", err)
+	}
+	return packet, nil
+}
+
+type sendFn func(addr mino.Address, msg serde.Message) error
+
+func doSend(addrs []mino.Address, msg serde.Message, send sendFn,
+	logger zerolog.Logger) chan error {
+	errs := make(chan error, len(addrs))
+	var wg sync.WaitGroup
+	wg.Add(len(addrs))
+	for _, addr := range addrs {
+		go func(addr mino.Address) {
+			defer wg.Done()
+			err := send(addr, msg)
+			if err != nil {
+				errs <- xerrors.Errorf("could not send to %v: %v", addr, err)
+				logger.Warn().Err(err).Msgf("could not send %T to %v", msg, addr)
+				return
+			}
+			logger.Debug().Msgf("sent %T to %v", msg, addr)
+		}(addr)
+	}
+
+	go func() {
+		wg.Wait()
+		close(errs)
+	}()
+	return errs
+}
+
+type unpackFn func(packet Packet) (mino.Address, serde.Message, error)
```

Comment:
we shouldn't have to use something like this - again, with proper serde usage.
see https://dedis.github.io/dela/#/serde